### ISSUE TITLE
[Feature Request] Make sure multiple concurrent speculative tasks aren't discarded

### ISSUE THREAD

**Issue by Sushisource** [temporalio/sdk-core#633]
Apparently server might send multiple speculative WFTs at once if there are concurrent updates. If it does, make sure that works.

**Comment by Quinn-With-Two-Ns** [comment]
To clarify the spec, I believe that if multiple concurrent speculative tasks are sent by the server it will end up discarding all but the last one. It is important on the SDK  side we don't let these concurrent tasks effect each other by say running in parallel or corrupting  the cached state of a workflow execution

**Comment by Sushisource** [comment]
> To clarify the spec, I believe that if multiple concurrent speculative tasks are sent by the server it will end up discarding all but the last one. It is important on the SDK side we don't let these concurrent tasks effect each other by say running in parallel or corrupting the cached state of a workflow execution

Riight, ok I remember our conversation better now. That definitely doesn't happen and isn't an issue.

So, what this really says is, like normal WFT, there is not any reason for me to buffer more than one of these at a time. Sounds like I'll need a 1-size buffer for non-query-only WFTs, and a longer buffer for query-only WFTS

**Comment by alexshtin** [comment]
Server might have only (no more than) one scheduled/started WFT at any given moment of time. It doesn't matter if it is normal or speculative. If update comes while WFT is running on the worker it will be sort of buffered on server and waiting for WT to complete. 

There is some sort of edge case though. If shard was moved (or pod restarted) during speculative WFT execution, server will lost information about it. And if new update comes in, server will create new WFT and will dispatch to the worker. Worker will get WFT with the same ScheduledId. We added protection on server and it now compares also timestamps of WFT. So when first WFT will be responded back to server as Completed, server will reject (return `NotFound`) because actual running WFT (from server prospective) has another timestamp.
